notes/pointer-basics.md

```diff
@@ -0,0 +1,219 @@
+
+We can use the **dereference operator (*)** to get the object to which the pointer points
+
+```C++
+int x = 333;
+int *p = &x; // p points to x
+std::cout << *p << std::endl; // prints 333
+
+```
+
+This means we can also assign to the object a pointer points to
+- Only if the pointer is not `const` though
+
+```C++
+int x = 333;
+int *p = &x;
+
+*p = 124;
+std::cout << *p << std::endl; // prints 124 now
+```
+
+---
+## Null pointers
+
+Null pointers do not point to any object
+- I.e. they point "nowhere"
+
+Dereferencing a null pointer will cause a segfault
+
+To get a null pointer:
+```C++
+int *p = 0;
+int *p2 = nullptr;
+int *p3 = NULL;
+// All 3 ways of initializing null pointer are basically equivalent
+
+```
+- C++ introduced `nullptr`, which has type `std::nullptr_t`
+- C's `NULL` macro is defined by `cstdlib` as `0`
+- However, `nullptr` avoids some ambiguity when resolving function overloads since `NULL` tends to get treated as an `int`
+
+---
+## Pointer arithmetic
+
+You can perform arithmetic on pointers, for example:
+```C++
+char *p = ...
+int *q = ...
+long *r = ...
+
+// Note: the "..." isn't valid C code, just filler
+// For the purposes of the example, let's assume:
+// p has value = 1000, q = 2000, r = 3000
+
+p += 3;
+q += 3;
+r += 3;
+
+// What are the values of p, q, and r now?
+
+```
+
+Answer:
+- `p = 1003, q = 2012, r = 3024`
+
+
+Why?
+
+First, the type of the object a pointer points to matters. In the above example, `q` points to an int object. `sizeof(int) == 4` on a 64-bit machine
+
+You might've already guessed the pattern. In general, for `typeName* ptrName`, the expression `ptrName + k` really evalutes to `ptrName + k * sizeof(typeName)`
+
+So in the above example `r + 3` would be the same as `r + 3*sizeof(long) = r + 3*8 = 3000 + 24 = 3024`
+
+
+---
+## void* Pointer
+
+The type `void*` is a special pointer type
+
+Pointer variables of this type can hold the address of any object
+- Basically the pointer holds an address, but we don't know the type of object at that address
+
+```C++
+int x = 333;
+void *p = &x;
+
+// Now p holds the address of x
+// You could cast p to be an int pointer, then dereference it to get x's value
+std::cout << *((int*)p) << std::endl;
+
+```
+
+
+---
+Now let's examine some common use cases for pointers
+
+---
+## Output parameters
+
+Remember that C/C++ passes arguments by value
+- Thus the callee receives a local copy of the argument
+- If the callee modifies a parameter, the caller's copy is NOT modified
+
+Bugged example:
+```C++
+void foo(int x){
+  x = 333;
+}
+
+int main(){
+  int x = 351;
+  foo(x);
+  std::cout << x << std::endl; // what gets printed?
+}
+
+```
+- Answer: `351` is printed! Updating `x` in function `foo` does not modify the caller (`main`'s) version of `x`
+
+Solution: we can mimic pass-by-reference using pointers!
+- Sure, the pointer VALUE (some address of an object) is passed by copy, but the address still indirectly gives access to the same object
+
```

Comment:
Although the use of the reference/dereference operator works for pass by reference, it is worth noting that `&` can be used too, in fact, it can be better. The reason I say this is because specifically for arrays, when you modify an array within another function, the pointer you passed into the function doesn't update its location whereas the local pointer used for the array in the function does. This is something that LOST actually battles with, and there are 2 ways to work around it. However, I think using `&` also is standard since it is not as ambiguous with `*`, since it can also denote an array.